File: src/main/java/com/vc/medium/L983.scala
package com.vc.medium

object L983 {
  def mincostTickets(days: Array[Int], costs: Array[Int]): Int = {
    val daysIncluded = new Array[Boolean](366)
    days.foreach(x => {
      daysIncluded(x) = true
    })

    val minCost = new Array[Int](366)
    (1 to 365).foreach(i => {
      if(daysIncluded(i)) {
        minCost(i) = minCost(i - 1) + costs(0)
        val minus7Days = Math.max(i - 7, 0)
        minCost(i) = Math.min(minCost(i), minCost(minus7Days) + costs(1))
        val minus30Days = Math.max(i - 30, 0)
        minCost(i) = Math.min(minCost(i), minCost(minus30Days) + costs(2))
      }
      else {
        minCost(i) = minCost(i - 1)
      }
    })
    minCost(365)
  }
}

File: src/main/java/com/vc/medium/L983.java
package com.vc.medium;

import java.util.*;

class L983 {
    public int mincostTickets(int[] days, int[] costs) {
        HashSet<Integer> daySet = new HashSet<>();
        for(int i = 0; i < days.length; i++) daySet.add(days[i]);

        int[] dp = new int[366];
        for(int i = 1; i < 366; i++) {
            if(daySet.contains(i)) {
                dp[i] = dp[i - 1] + costs[0];

                int day7DayAgo = Math.max(i - 7, 0);
                dp[i] = Math.min(dp[i], dp[day7DayAgo] + costs[1]);

                int day30DayAgo = Math.max(i - 30, 0);
                dp[i] = Math.min(dp[i], dp[day30DayAgo] + costs[2]);
            }
            else {
                dp[i] = dp[i - 1];
            }
        }
        return dp[365];
    }
}
